Fix DaemonSet cache mutation

ncdc opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md and developer guide https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**: stops the DaemonSetController from mutating the DaemonSet shared informer cache

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #38985

cc @deads2k @mikedanese @lavalamp @smarterclayton

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/38986)
<!-- Reviewable:end -->

ncdc commented:
Working on fixing the unit test failure.

smarterclayton commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -539,19 +540,26 @@ func storeDaemonSetStatus(dsClient unversionedextensions.DaemonSetInterface, ds
 		return nil
 	}
 
+	clone, err := conversion.NewCloner().DeepCopy(ds)
```
Is this the new convention for invoking deep copy?  I thought we were using the copy method directly now?

ncdc replied:
I saw this somewhere in the code. What is the correct way to do it now?

smarterclayton replied:
Majority of instances I see are doing api.Scheme.DeepCopy()

ncdc commented:
My changes make the unit tests flakey. Investigating.

Commit:
Fix DaemonSet controller cache mutation

Add dsStoreSynced so we also wait on this cache when starting the
DaemonSetController.

Switch to using a fake clientset in the unit tests.

Fix TestNumberReadyStatus so it doesn't expect the cache to be mutated.

mikedanese commented:
cc @kubernetes/daemonset

ncdc commented:
OK, I compiled the unit tests for this package and ran them in an infinite loop for a few minutes and didn't get any failures. I'd appreciate it if someone else could try the same and let me know if you see any flakes.

lukaszo commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/38986)*
<!-- Sent from Reviewable.io -->

lukaszo commented:
I've run DaemonSet unit tests in loop 100 times and all runs passed.

kargakis commented:
LGTM. Was the test failure related to https://github.com/kubernetes/kubernetes/issues/38518?

ncdc commented:
@kargakis no, it was because the shared informer factory was being started in the unit tests, which makes things a bit unpredictable
